src/swe/swe_operators.c

```diff
@@ -0,0 +1,364 @@
+#include "swe_operators.h"
+
+#include "swe_operators_impl.h"
+
+// frees a data context allocated using PETSc, returning a libCEED error code
+static int FreeContextPetsc(void *data) {
+  if (PetscFree(data)) return CeedError(NULL, CEED_ERROR_ACCESS, "PetscFree failed");
+  return CEED_ERROR_SUCCESS;
+}
+
+// creates a QFunction context for a flux or source operator with the given
+// minimum water height threshold
+static PetscErrorCode CreateQFunctionContext(Ceed ceed, PetscReal tiny_h, CeedQFunctionContext *qf_context) {
+  PetscFunctionBeginUser;
+
+  SWEContext swe_ctx;
+  PetscCall(PetscCalloc1(1, &swe_ctx));
+
+  swe_ctx->dtime   = 0.0;
+  swe_ctx->tiny_h  = tiny_h;
+  swe_ctx->gravity = 9.806;
+
+  CeedQFunctionContextCreate(ceed, qf_context);
+  CeedQFunctionContextSetData(*qf_context, CEED_MEM_HOST, CEED_USE_POINTER, sizeof(*swe_ctx), swe_ctx);
+  CeedQFunctionContextSetDataDestroy(*qf_context, CEED_MEM_HOST, FreeContextPetsc);
+  CeedQFunctionContextRegisterDouble(*qf_context, "time step", offsetof(struct SWEContext_, dtime), 1, "Time step of TS");
+  CeedQFunctionContextRegisterDouble(*qf_context, "small h value", offsetof(struct SWEContext_, tiny_h), 1,
+                                     "Height threshold below which dry condition is assumed");
+  CeedQFunctionContextRegisterDouble(*qf_context, "gravity", offsetof(struct SWEContext_, gravity), 1, "Accelaration due to gravity");
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode CreateSWEFluxOperator(Ceed ceed, RDyMesh *mesh, int num_boundaries, RDyBoundary boundaries[num_boundaries],
```

Comment:
How about splitting this long function into two (e.g., `CreateSubOpForInternalEdges` and `CreateSubOpForBoundaryEdges`)?

Reply:
That's a good idea. I'll break it up.